File: 1161-maximum-level-sum-of-a-binary-tree/golang/bfs.go

import (
	"math"
)

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type bfsItem struct {
    root *TreeNode
    level int
}

func maxLevelSum(root *TreeNode) int {
    if root == nil {
        return 0
    }

    queue := []bfsItem{{
        root: root,
        level: 1,
    }}


    max := math.MinInt
    max_l := 1
    currLevel := 1
    sum := root.Val

    for len(queue) > 0 {
        curr, l := queue[0].root, queue[0].level
        queue = queue[1:]
        if currLevel != l {
            if  sum > max {
                max = sum
                max_l = currLevel
            }
            currLevel++
            sum = 0
        }

        sum += curr.Val

        if curr.Left != nil {
            queue = append(queue, bfsItem{
                root: curr.Left,
                level: l+1,
            })
        }
        if curr.Right != nil {
            queue = append(queue, bfsItem{
                root: curr.Right,
                level: l+1,
            })
        }
    }

    if sum > max {
        return currLevel
    }

    return max_l
}

File: 1161-maximum-level-sum-of-a-binary-tree/java/bfs.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    class Pair {
        public int level;
        public TreeNode node;
        public Pair(int level, TreeNode node) {
            this.level = level;
            this.node = node;
        }
    }
    
    public int maxLevelSum(TreeNode root) {
        Queue<Pair> bfs = new LinkedList<Pair>();
        
        int max = Integer.MIN_VALUE;
        int levelMax = -1;
        
        int currentSum = 0;
        int currentLevel = 1;
        
        bfs.add(new Pair(1, root));
        
        while(!bfs.isEmpty()) {
            Pair tmp = bfs.remove();
            
            if(currentLevel != tmp.level) {
                if(currentSum > max) {
                    max = currentSum;
                    levelMax =  currentLevel;
                }
                currentSum = 0;
                currentLevel = tmp.level;
            }
            
            currentSum += tmp.node.val;
            
            if(tmp.node.left != null) {
                bfs.add(new Pair(tmp.level + 1, tmp.node.left));
            }
            if(tmp.node.right != null) {
                bfs.add(new Pair(tmp.level + 1, tmp.node.right));
            }
        }
        
        if(currentSum > max) {
            levelMax = currentLevel;
            max = currentSum;
        }
        
        return levelMax;
    }
}